### PR TITLE
Basic splitting of TorchToLinalg.cpp.

### PR DESCRIPTION
**Pull request by pashu123**
This is to address https://github.com/llvm/torch-mlir/issues/382
Before this file gets even larger let's split it.

1. I am yet to write all the common functions under a namespace.
2. All the common functions should go to `Utils.h` under utils dir or `CommonTorchToLinalg.h` like we have in this PR?

### CODE REVIEW

**Comment by pashu123** [comment]
@cathyzhyi I have started to split files in torchToLinalg.cpp. Initial suggestion needed.

**Comment by pashu123** [comment]
@ramiro050 Any suggestions?

**Comment by cathyzhyi** [comment]
This is a really good starting point! Thanks for making the changes.
Can you refer to the original issue in the comment so that people not familiar with this would know the background. 

> I am yet to write all the common functions under a namespace

yea, we will need to put them under torch namespace.

> All the common functions should go to Utils.h under utils dir or CommonTorchToLinalg.h like we have in this PR?

I think `CommonTorchToLinalg.h` as it is is fine.

Is the current plan using one .cpp file for just one pattern? I think we can probably just put all math related ops into one file and use `populateLoweringMathOpPattern` to populate all of them together. 

@silvasean Do you have any suggestions on how to split the file?

**Comment by pashu123** [comment]
> This is a really good starting point! Thanks for making the changes. Can you refer to the original issue in the comment so that people not familiar with this would know the background.
> 
> > I am yet to write all the common functions under a namespace
> 
> yea, we will need to put them under torch namespace.
> 
> > All the common functions should go to Utils.h under utils dir or CommonTorchToLinalg.h like we have in this PR?
> 
> I think `CommonTorchToLinalg.h` as it is is fine.
> 
> Is the current plan using one .cpp file for just one pattern? I think we can probably just put all math related ops into one file and use `populateLoweringMathOpPattern` to populate all of them together.
> 
> @silvasean Do you have any suggestions on how to split the file?

@cathyzhyi Yes, we put all math-related ops into one file and use `populateLoweringMathOp..` as suggested. We can have elementwise, math, reduction etc as different categories. We can pull out ops from torchTolinalg and put them into this scheme incrementally. The original issue is here: https://github.com/llvm/torch-mlir/issues/382

File: include/torch-mlir/Conversion/TorchToLinalg/CommonTorchToLinalg.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
+#ifndef TORCH_COMMON_MLIR_CONVERSION_ATENTOLINALG_ATENTOLINALG_H
```
Comment (cathyzhyi) [inline include/torch-mlir/Conversion/TorchToLinalg/CommonTorchToLinalg.h:1]: We need license headers for all the new files.

**Review by cathyzhyi — CHANGES_REQUESTED** [review]

**Comment by pashu123** [comment]
@silvasean: I and @cathyzhyi had a small discussion yesterday and we have two possible layouts:-
1. Put every op lowering in one file under generalized dir: For eg., Under the `view` directory we can have transpose.cpp, permute.cpp, reshape.cpp files containing respective op lowerings.
2. Put ops that share a similar logic in a single file under a generalized dir, for eg., transpose.cpp and permute.cpp share a similar logic and hence their lowering can be put into a single cpp file inside `view` dir. Similarly, aten::mat.mul, aten::linear, aten::bmm op lowering can be put into a single cpp file under `math` or `reduce` dir.

**Comment by silvasean** [comment]
I would prefer option 2. Option 1 is a bit too fine-grained, and is not well-defined when multiple ops share lots of common functionality.

We should break it into coarse categories, that are easy for people to guess which file a particular op is in.

I gave a suggested splitting in https://github.com/llvm/torch-mlir/issues/382#issue-1036924035, but I'm not attached to that. Anything that seems reasonable is fine for now. This is mostly about code sharing, without being too confusing.

Thanks for doing this :)

**Comment by Shukla-Gaurav** [comment]
This is now covered by https://github.com/llvm/torch-mlir/pull/657. @pashu123 I think we should close this.

**Pull request closed without merging**